Improve RArrowDS and introduce RCombinedDS

ktf opened the pull request.

ktf committed:
Improve RArrowDS

* Provide hints of what branches are likely and what not.
* Speedup inner loop by moving some checks one level up in the outer
  loop.
* Actually support rewinding an RArrowDS.

ktf committed:
Introduce RCombinedDS to combine other two datasources

This introduces a proxy RCombinedDS which can be used to combine entries
of other two datasources, effectively mimicking various kind of JOIN
operations between the two.

The actual index in implemented by composition using a RCombinedDSIndex
derived class.

This also exposes the issue mentioned in
https://github.com/root-project/root/pull/3433 and
https://github.com/root-project/root/pull/3428 where the cursor for
the RArrowDS is not updated correctly in case of SetEntry moving
backwards (like it is needed for a double loop).

pcanal commented on tree/dataframe/src/RCombinedDS.cxx:
```diff
@@ -0,0 +1,195 @@
+ *************************************************************************/
+
+// clang-format off
+/** \class ROOT::RDF::RCombinedDS
```
Is there a more precise name we could use?  Combined is a bit too generic (does not evoke Cartesian product).

ktf replied:
Actually I should update the documentation, since this now allows any kind of combination between the rows of the two combined datasources, not limiting it to the cross product. The actual logic for how to generate the couples of rows is provided by a `RCombinedDSIndex` derived class. So this is effectively combining the rows of two datasources in a generic way and `RCombinedDS` seems to be a good choice to me. I am of course open to other suggestions. Do you have anything in mind?

pcanal replied:
> since this now allows any kind of combination between the rows

I completely missed that.  So indeed the generality of the name is fine ... except that it 'defaults' silently to CrossProduct.

>  So this is effectively combining the rows of two datasources in a generic way

Since this is the case and the intent (this is a good thing), I would definitively add another example (for example the flat/friend combinations).

And then back to the default, when reading I assumed a default of 'flat/friend' combinations, so I would make that the default .... which then brings in the questions of the interface to change the combination index which currently seems to be of the style:
```
auto combined = std::make_unique<RCombinedDS>(std::move(left), std::move(right), std::make_unique<RCombindedDSCrossJoinIndex>() );
```
which seems a bit verbose, any idea on simplifying it or is that good enough?

ktf replied:
Ok, I will clarify the documentation, implement the friend / `UNION`-like index and make that the default. Regarding your last question, I am not particularly happy either and I am trying to reduce the boilerplate. To start with I was thinking of having some helper:

```C++
template <typename INDEX>
std::unique_ptr<RCombinedDS> Combine<INDEX>(std::unique_ptr<RDataSource>, std::unique_ptr<RDataSource>, std::string left_prefix = "left_", std::string right_prefix = "right_") {...}
```

but suggestions are welcome.

Another alternative I was thinking about would be to introduce a `template <typename CONCRETE_DATASOURCE> RDataSourceNode` class which removes the need for the `#define protected public` and should allow removing the unique_ptr as well (because the RDataSourceNode would really be treated as a value). So it will become:

```C++
template <typename INDEX, typename LEFT, typename RIGHT>
std::unique_ptr<RCombinedDS> Combine<INDEX>(RDataSourceNode<LEFT>, RDataSourceNode<RIGHT>, std::string left_prefix = "left_", std::string right_prefix = "right_") {...}
```

but I am not sure that will result in less verbose code on the client side...

pcanal replied:
@dpiparo Any suggestions?

ktf commented:
I've updated the PR with todays developments. Apart from the changes requested I've introduced a new index which allows performing all sort of block diagonal combinations between the indices of two tables, where each block represent a category of objects, e.g. tracks belonging to the same event. This is still not fully tested, but it should give an idea of the possibilities.

Axel-Naumann commented:
OK let's get this merged. We need two things:
- find a better name than "combined". I suggest `RCrossJoinDS` or `RCartesianProductDS`.
- fix the conflict. I can certainly do that myself.

@ktf let me know, please!

ktf commented:
Notice this is not limited to provide Cross Join / Cartesian Products kind of joins (i.e. MxN) but it's fully configurable via the index, that's why I chose "Combined". In principle one could have an index which repeats ten times the same pair of "rows" or whatever. RAssociativeDS?

ktf commented:
See also thread we had with @pcanal at https://github.com/root-project/root/pull/3443/files#r258137063

eguiraud commented:
Hi, I'm picking this up (thanks @linev for assigning me).

#3433 should be merged first. It just requires rebasing and fixing a trivial conflict.

For this PR: I will review the code one last time before approving, mostly to be familiar with the code myself. I don't expect any issue. I think `RCombinedDS` is not a bad name given that it really provides arbitrary combinations of datasources. What do you think @Axel-Naumann ? @ktf proposed `RAssociativeDS` as an alternative but I don't think it's clearer.

eguiraud commented:
Discussing with @ktf, we decided to add `RCombinedDS` as a friend of `RDataSource` to avoid the `#define protected public` that is now required to call `GetColumnReadersImpl` from `RCombinedDS`. I will do that and resolve the few conflicts this PR now has before merging.

I don't think the other comments still apply (right @pcanal ?)

eguiraud commented:
...thank you @phsft-bot

pcanal reviewed (changes requested):
need conflict resolution.

eguiraud commented:
@pcanal also needs a couple more things, see my last comment. It's on my to do list.

guitargeek commented:
@ktf, what should we do with this PR? Should we closed it?

ktf commented:
Up to you what you want to do. We have decided to go a different way than RDataFrame for our analysis framework, so it's not critical for us. Still I think some way of combining RDataSources with JOIN like operations is probably something useful.

guitargeek commented:
Ok, thanks for the clarification! That's true, the `RCombinedDS` is nice. I guess the RDataFrame devs should take your PR from here. But right now we don't enable `arrow` in the CI. Let's first do that @martamaja10 @vepadulano @dpiparo

guitargeek commented:
@ktf, thanks for the activity here! I have one question though: if you say you're going a different route other than RDataFrame for your analysis framework, do you still need the RArrowDS? As far as I can tell, it's a "hidden" ROOT feature that you only get when building with `arrow=ON`, which we don't test in any of our CI runs. So I'm not even sure if it works at this point. Is it worth to keep it, if you were the author and only user, but you have now moved on? ROOT classes that require extra dependencies should not be kept without reasons I think.

FYI, @dpiparo
